### PR TITLE
New method for calculating grain volumes

### PR DESCRIPTION
**Pull request by permcody**
This PR adds a new class for calculating grain volumes using integrals instead of just summing whole element volumes. This replaces the need to output CSV from the FeatureFloodCount object.

@frombs - You'll want to give this a whirl with your study and see if it helps. Make sure that your execute_on matches this object and that you turn "compute_op_maps" on in the GrainTracker. If you have any questions let me know.

I will submit a follow-on PR to deprecate the existing volume calculation method and start removing it from all of the existing tests. 

@SudiptaBiswas - This new VectorPostprocessor may be of interest to you. If you want to use it instead of getting volumes from the GrainTracker, we should chat. I could add the equivalent interface to this new object as you have now. It should be a lot more accurate.

### CODE REVIEW

**Comment by SudiptaBiswas** [comment]
Looks like `FeatureVolumeVectorPostprocessor` calculates volume in a similar pattern as `ComputeGrainCenterUserObject` but in a more flexible `GrainTracker` way. I would certainly like to use this.  It would be good to have a `getGrainVolume` type of API I think.

**Comment by permcody** [comment]
Ok - I'll get that added once Daniel takes a look at this. The GrainCenterUO is the one we are eventually going to remove correct?

**Comment by SudiptaBiswas** [comment]
Yes, it's been deprecated already.

**Comment by permcody** [comment]
... And I just added that API. It was a small two line addition so it's in there.

File: modules/phase_field/src/base/PhaseFieldApp.C [hunk 1]
```diff
@@ -236,10 +236,10 @@
...
  * VectorPostprocessors
  */
- #include "GrainCentersPostprocessor.h"
- #include "GrainForcesPostprocessor.h"
- #include "GrainTextureVectorPostprocessor.h"
-
+#include "GrainCentersPostprocessor.h"
+#include "GrainForcesPostprocessor.h"
+#include "GrainTextureVectorPostprocessor.h"
+#include "FeatureVolumeVectorPostprocessor.h"
```
Comment (dschwen) [inline modules/phase_field/src/base/PhaseFieldApp.C]: We order those (and the register calls alphabetically - except deprecated registrations, which are at the end of each block)

File: modules/phase_field/src/vectorpostprocessors/FeatureVolumeVectorPostprocessor.C [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+{
+  const GrainTrackerInterface * grain_tracker = dynamic_cast<const GrainTrackerInterface *>(&_flood_counter);
+
+  if (!grain_tracker)
+    mooseError("FeatureVolumeVectorPostprocessor currently only supports the GrainTrackerInterface");
+
+  const auto num_grains = grain_tracker->getNumberGrains();
+  _grain_volumes.assign(num_grains, 0);
+
+//  const FeatureFloodCount::FieldType field_type = FeatureFloodCount::FieldType::UNIQUE_REGION;
```
Comment (dschwen) [inline modules/phase_field/src/vectorpostprocessors/FeatureVolumeVectorPostprocessor.C]: This can be removed?

**Comment by frombs** [comment]
@permcody - I have been using execute_on = 'initial timestep_end' but you used only "initial" in your test file.  Is "timestep_end" still okay to use?

**Comment by permcody** [comment]
Yes. That's just because this test is to verify the initial condition. It's
a steady test

**Comment by permcody** [comment]
I'm not sure I like what Github is doing here with updates. I pushed another commit to resolve the comments on your review. You can verify that by looking at the latest commit. Your comments aren't showing as updated though.

**Comment by frombs** [comment]
When I run my test case I get an inconsistent number of grain volumes being output to the CVS files at each time step.  For example, time step 0 reports 40 rows of data, time steps 1 - 5 contain 38 rows, and time step 6-10 contain 36 rows.

**Comment by permcody** [comment]
Oh right... I keep forgetting you need reports for inactive grains. I can
fix that.

**Comment by permcody** [comment]
@frombs - The printout of the volumes from the vector postprocessor should be "stable" It should only grow (when new grains are nucleated - which doesn't effect you), not shrink. Give this branch a whirl.

**Review by dschwen — APPROVED** [review]

**Comment by dschwen** [comment]
Are we waiting for @frombs' feedback here?

**Comment by permcody** [comment]
Nah, this work is in progress. The sooner we get this in, the sooner I can
start deprecating the existing method. I should be able to remove quite a
bit of code from FeatureFloodCount and GrainTracker. If there are bugs,
I'll supply more patches but this is pretty straightforward if you look at
the logic.

**Comment by frombs** [comment]
@permcody - How do I control the threshold for the new method?

**Comment by permcody** [comment]
The new method takes the integral of the solution value on ALL elements identified in the original flood, (i.e. the elements identified by `threshold` and `connecting_threshold`). As a result you'll get a fairly accurate solution that's only off due to discretization (the finite elements) and truncation (cutting off contributions from elements with a lower value than your threshold) error.

I plan to deprecate that other newer volume threshold and the existing volume file stuff out of those objects. I'd rather only support this one new method since it should be far more accurate. I added a new test to compute the area of a few smooth bounding boxes and smooth circle the result is accurate to ~3 decimal places (I'm not sure if I should expect a closer answer than that since I haven't really looked at the expected volume of those diffuse interface objects).

**Comment by frombs** [comment]
Below is a table comparing results of the exact solution with the old and new methods.  The new method is more consistent in that it added the individual contributions from the split grains.  The old method did not include both contributions (see values in red).  However, except for the split grain issue, the values from the old method agreed better with the exact solution for time step 0. The new integration method slightly under-predicted the areas for time step 0.  Also, at time step 2 many of the large grains actually see an increase in value before decreasing again in step 3.  Any ideas to explain that behavior?

volume_comparison

**Comment by frombs** [comment]
Below is a plot of the new method showing the dip in area before increasing again.  The dashed line is the average grain area.
default_threshold

**Comment by permcody** [comment]
We may want to have @tonkmr or @dschwen chime in here but I don't think we
should consider the area of grains in the initial IC. It's a poor
approximation since the interface is not diffuse. I think it's completely
defensible to just leave it out.

**Comment by frombs** [comment]
I can live with that.  However, I need to understand why we get the increase in step 2 and then a decrease at step 3 for the big grains so I can explain that in a paper.  If the drop between step 1 and 0 can be explained I am okay with it too.

**Comment by permcody** [comment]
I wonder if the interface is all set up at that point. I'm not a phase-field expert but I typically run fast and loose with the interface, barely resolving it, barely giving it enough resolution. I'm not sure what you are doing in your sim, but I'd suggest making sure you have plenty of adaptivity and that grain boundary width and adaptivity are within the right proportions. 

I just thought of another problem. You are doing with with EBSD data and those split grains. Right now I'm outputting the volumes per grain index which is probably not right. You can have a single EBSD grain assigned to different indices... I guess we need to dump this information per "feature id" instead.

**Comment by permcody** [comment]
Now that this is a vector postprocessor, I could actually just dump a second vector with the feature_ids then you could interpret the data like you want. Either way is fine, it's an easy change.

**Comment by permcody** [comment]
@frombs - I also welcome you to stop by the cube at some point. We might be able to have a face to face to talk through a lot of these issues.

**Comment by frombs** [comment]
Would the second vector contain the feature_id per element?  I will be at MFC Friday doing experiments but can meet on Monday if you have time.

**Comment by permcody** [comment]
No, the feature_id would line up with the list of grain volumes. You could
put them together to track volume per feature_id.

**Pull request merged**